### DISCUSSION TITLE
very weird perf issue

### DISCUSSION THREAD

**Discussion by omgoshjosh** [react-hook-form/react-hook-form#3442] [Q&A]
Hello again,

I have started seeing some weird performance degradations, but only when the dev tools are open and MOST strangely, when the chrome performance tool IS NOT CAPTURING. That is, when I run the chrome performance profiler, when capturing (or when dev tools are not open at all), the performance is totally fine.

Below is the component that is using `Controller` and below that is what the REACT profiler is showing me (~300 ms per rerender of `Controller` which seems like a lot. DpSwitch is just a wrapper around this library https://github.com/markusenglund/react-switch. I'm really just looking for some debugging tips to see if i can figure out why that controller is taking 300ms to render. toggles is a constant with 5 elements so the map shouldn't be a problem. I am having trouble seeing inside the Controller and why it's rerendering that switch so many times. (there are 25 total to be rendered and they are getting rerendered a lot)

```tsx
export function InviteeToggles({
    control,
    name,
    toggleData,
}: IInviteeTogglesProps) {
    return (
        <Controller
            control={control}
            name={name}
            defaultValue={toggleData}
            render={({ onChange, value }) => (
                <div className="dpm-flx dpm-flx-cell-1">
                    {toggles.map((toggle: Toggle, idx) =>
                        <div key={idx} className="dpm-flx-cell-1 dpm-pd-2-left sg-typ-center">
                            <DpSwitch
                                className="sg-m-iblock md-primary dpm-mg-0-enforce"
                                ariaLabel={`${toggle.title} switch`}
                                checked={value[toggle.id]}
                                onChange={(checked) => {
                                    const newToggles = onToggleAccessOption(
                                        checked,
                                        idx,
                                        value,
                                    );
                                    onChange(newToggles);
                                }}
                                disabled={isToggleDisabled(
                                    value,
                                    idx,
                                )}
                            />
                        </div>
                    )}
                </div>
            )}
        />
    );
}
```

![image](https://user-images.githubusercontent.com/1960929/99007712-9deddc00-24f9-11eb-85eb-e59c2eb86f6c.png)

**Comment by omgoshjosh** [comment]
it's possible it's that rgba calculation on the boxshadow? it just seems weird that it would take so long just to render these 25 switches

```tsx
export function DpSwitch({
    checked,
    onChange,
    className,
    ariaLabel,
    disabled = false,
    onColor = "#99a2d5",
    onHandleColor = "#3f51b5",
    handleDiameter = 20,
    height = 14,
    width = 34,
    stopPropagation = false,
}: IDpSwitchProps) {
    return (
        <div onClick={e => stopPropagation && e.stopPropagation()} style={{ marginRight: '8px' }}>
            <Switch
                className={className}
                checked={checked}
                disabled={disabled}
                onChange={onChange}
                onColor={onColor}
                onHandleColor={onHandleColor}
                handleDiameter={handleDiameter}
                uncheckedIcon={false}
                checkedIcon={false}
                boxShadow="0px 1px 5px rgba(0, 0, 0, 0.6)"
                activeBoxShadow="0px 0px 1px 10px rgba(0, 0, 0, 0.2)"
                height={height}
                width={width}
                aria-label={ariaLabel}
            />
        </div>
    );
}
```

**Reply by bluebill1049** [reply]
wow that's a lot of re-render for a simple Controller component, do you have a CSB? worth to digging into the reason why, it should isolate the re-render at least not impact your entire app/form. still 300ms is crap

**Reply by bluebill1049** [reply]
a simple CSB would give me good context.

**Reply by omgoshjosh** [reply]
ok i got it, the links themselves are self explanatory by name, but if you have any other questions i can try to answer them. thanks again.

[Controller on each switch](https://codesandbox.io/s/controller-on-each-switch-ltxet)

[One Controller for each switch](https://codesandbox.io/s/one-controller-for-all-switches-kwr06)

**Reply by omgoshjosh** [reply]
yeah it looks like even if you just change one input with one letter, it triggers a rerender on all the switches.

**Reply by omgoshjosh** [reply]
now it seems to be gone. i am so tired lol. anyway, my boss at work thinks it may be related to recent weird stuff happening with apple.

https://twitter.com/lapcatsoftware/status/1326990296412991489

https://www.macrumors.com/2020/11/12/apple-card-apple-pay-maps-outages/

etc

**Reply by omgoshjosh** [reply]
and its back lol. jesus lord have mercy.
